src/webgpu/util/check_contents.ts

```diff
@@ -0,0 +1,234 @@
+import {
+  assert,
+  ErrorWithExtra,
+  iterRange,
+  range,
+  TypedArrayBufferView,
+  TypedArrayBufferViewConstructor,
+} from '../../common/util/util.js';
+
+export type BufferCheckGenerator = (index: number) => number;
+export type BufferCheckPredicate = (index: number, value: number) => boolean;
+export type BufferCheckPredicatePrinter = {
+  leftHeader: string;
+  cell: (index: number) => number;
+}[];
+
+/**
+ * Check whether two `TypedArray`s have equal contents.
+ * Returns `undefined` if the check passes, or an `Error` if not.
+ */
+export function checkElementsEqual(
+  actual: TypedArrayBufferView,
+  expected: TypedArrayBufferView
+): ErrorWithExtra | undefined {
+  assert(actual.constructor === expected.constructor, 'TypedArray type mismatch');
+  assert(actual.length === expected.length, 'size mismatch');
+  return checkElementsEqualGenerated(actual, i => expected[i]);
+}
+
+/**
+ * Check whether each value in a `TypedArray` is between the two corresponding "expected" values
+ * (either `a[i] <= actual[i] <= b[i]` or `a[i] >= actual[i] => b[i]`).
+ */
+export function checkElementsBetween(
+  actual: TypedArrayBufferView,
+  expected: readonly [TypedArrayBufferView, TypedArrayBufferView]
+): ErrorWithExtra | undefined {
+  const e = checkElementsPassPredicate(
+    actual,
+    (index, value) =>
+      value >= Math.min(expected[0][index], expected[1][index]) &&
+      value <= Math.max(expected[0][index], expected[1][index]),
+    {
+      predicatePrinter: [
+        { leftHeader: 'between', cell: index => expected[0][index] },
+        { leftHeader: 'and', cell: index => expected[1][index] },
+      ],
+    }
+  );
+  return e === undefined
```

Comment:
I'm not sure I understand what this does. Maybe it could be a method on ErrorWithExtra so we can do `e??.dostuff` the method name would help make what happens more explicit